### ISSUE TITLE
Rule proposal: prevent both children and dangerouslySetInnerHTML at the same time

### ISSUE THREAD

**Issue by lencioni** [yannickcr/eslint-plugin-react#710]
React will issue a warning if you try to use dangerouslySetInnerHTML and children at the same time.

https://github.com/facebook/react/blob/85dcbf83/src/renderers/dom/shared/ReactDOMComponent.js#L170

This should be easy enough to enforce for most cases with a linter rule.

Bad:

```jsx
<div dangerouslySetInnerHTML={{ __html: 'HTML' }}>
  Children
</div>
```

Good:

```jsx
<div dangerouslySetInnerHTML={{ __html: 'HTML' }} />
```

```jsx
<div>
  Children
</div>
```

**Comment by petersendidit** [comment]
I can work on this. What should this rule be named? `jsx-no-children-and-dangerouslySetInnerHTML` ?

**Comment by ljharb** [comment]
The rule needs to be all lowercase - maybe `jsx-no-danger-with-children`?

**Comment by lencioni** [comment]
Since this is scoped to DOM elements, perhaps that should be part of the name? In #709 we settled on `void-dom-elements-no-children`, so perhaps this should be `dom-elements-children-or-inner-html` or `dom-elements-no-inner-html-with-children` (or `dom-elements-no-danger-with-children`, but I worry that "danger" may end up being too broad at some point)?

I think that the `dangerouslySetInnerHTML` is a React thing, not a JSX thing, so I'm not sure this should come with the `jsx-` prefix.

**Comment by ljharb** [comment]
Ah yes, that's true, and it should catch `React.createElement` uses as well.

I think `dom-elements-no-danger-with-children` or `dom-elements-no-inner-html-with-children` is fine.

**Comment by lencioni** [comment]
Thinking about #712, which I think might be named `no-inner-html`, I think `dom-elements-no-danger-with-children` might be better actually. Plus it makes me laugh.

**Comment by petersendidit** [comment]
New commit should take care of the comments. Let me know if there is anything else.
